ADApp/ADSrc/NDArrayPool.cpp

```diff
@@ -121,156 +98,97 @@ NDArray* NDArrayPool::getNextFreeArray(NDArray *pArray)
-            closestArray = nextArray;
-          }
-        }
-        nextArray = getNextFreeArray(nextArray);
-      }
-      if (threshArray) {
-        pArray = threshArray;
-      } else {
-        if (closestArray) {
-          pArray = closestArray;
-        }
-      }
-    }
+    
+  /* Initialize fields */
+  pArray->pNDArrayPool = this;
+  pArray->referenceCount = 1;
+  pArray->pDriver = pDriver_;
+  pArray->dataType = dataType;
+  pArray->ndims = ndims;
+  memset(pArray->dims, 0, sizeof(pArray->dims));
+  for (int i=0; i<ndims && i<ND_ARRAY_MAX_DIMS; i++) {
+    pArray->dims[i].size = dims[i];
+    pArray->dims[i].offset = 0;
+    pArray->dims[i].binning = 1;
+    pArray->dims[i].reverse = 0;
   }
 
-  if (pArray) {
-    /* We have a frame */
-    /* Initialize fields */
-    pArray->pNDArrayPool = this;
-    pArray->pDriver = pDriver_;
-    pArray->dataType = dataType;
-    pArray->ndims = ndims;
-    memset(pArray->dims, 0, sizeof(pArray->dims));
-    for (i=0; i<ndims && i<ND_ARRAY_MAX_DIMS; i++) {
-      pArray->dims[i].size = dims[i];
-      pArray->dims[i].offset = 0;
-      pArray->dims[i].binning = 1;
-      pArray->dims[i].reverse = 0;
-    }
-    /* Erase the attributes if that global flag is set */
-    if (eraseNDAttributes) pArray->pAttributeList->clear();
-    // calcs totalBytes
-    pArray->getInfo(&arrayInfo);
-    if (dataSize == 0) dataSize = arrayInfo.totalBytes;
-    if (!freeArray) {
-      if (arrayInfo.totalBytes > dataSize) {
-        // we have a passed array, check size
-        printf("%s: ERROR: required size=%d passed size=%d is too small\n",
-        functionName, (int)arrayInfo.totalBytes, (int)dataSize);
-        pArray=NULL;
+  /* Erase the attributes if that global flag is set */
+  if (eraseNDAttributes) pArray->pAttributeList->clear();
+
+  /* At this point pArray exists, but pArray->pData may be NULL */
+  /* If the caller passed a valid buffer use that */
+  if (pData) {
+    pArray->pData = pData;
+  } else if (pArray->pData == NULL) {
+    if ((maxMemory_ > 0) && ((memorySize_ + dataSize) > maxMemory_)) {
+      // We don't have enough memory to allocate the array
+      // See if we can get memory by deleting arrays
+      // Delete the largest arrays first, i.e. work from the end of freeList_
+      NDArray *freeArray;
+      std::multiset<freeListElement>::iterator it;
+      while (!freeList_.empty() && ((memorySize_ + dataSize) > maxMemory_)) {
+        it = freeList_.end();
+        it--;
+        freeArray = it->pArray_;
+        freeList_.erase(it);
+        memorySize_ -= freeArray->dataSize;
+        numBuffers_--;
+        delete freeArray;
       }
     }
-  }
-
-  if (pArray) {
-    /* If the caller passed a valid buffer use that, trust that its allocated size is correct */
-    if (pData) {
-      pArray->pData = pData;
+    if ((maxMemory_ > 0) && ((memorySize_ + dataSize) > maxMemory_)) {
+      printf("%s: error: reached limit of %ld memory (%d buffers)\n",
```

Comment:
Should this be an asynPrintf call or removed?  Probably there is no asynUser available to pass to the asyn print calls...

Reply:
Same comment as above.